clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala

```diff
@@ -0,0 +1,244 @@
+      new ApiClient(conf.apiURL, conf.accessKey, conf.secretKey).getRangeURL("test", rangeID)
+    SSTableReader
+      .forRange(new Configuration(), location)
+      .newIterator()
+      .map(a =>
+        (
+          new String(a.key),
+          new String(a.message.address),
+          a.message.addressType.isRelative,
+          getSeconds(a.message.lastModified)
+        )
+      )
+      .toSet
+  }
+  def getAddressesTuples(
+      rangeIDs: Set[String],
+      conf: APIConfigurations
+  ): Set[(String, String, Boolean, Long)] = {
+    if (rangeIDs.isEmpty) Set()
+    else
+      getEntryTuples(rangeIDs.head, conf).union(getAddressesTuples(rangeIDs.tail, conf))
+  }
+
+  def subtractAddressesInRange(
+      addresses: Set[(String, String, Boolean, Long)],
+      rangeID: String,
+      conf: APIConfigurations,
+      repo: String
+  ): Set[(String, String, Boolean, Long)] = {
+    val activeAddresses = getRangeAddresses(rangeID, conf, repo)
+    addresses.filterNot(x => activeAddresses.contains(x._1))
+  }
+
+  /** @param leftRangeIDs
+   *  @param rightRangeIDs
+   *  @param conf
+   *  @return tuples of type (key, address, isRelative, lastModified) for every address existing in leftRanges and not in rightRanges
+   */
+  def leftAntiJoinAddresses(
+      leftRangeIDs: Set[String],
+      rightRangeIDs: Set[String],
+      conf: APIConfigurations,
+      repo: String
+  ): Set[(String, String, Boolean, Long)] = {
+    @tailrec
+    def subtractAddressesInRanges(
+        addresses: Set[(String, String, Boolean, Long)],
+        rangeIDs: Set[String]
+    ): Set[(String, String, Boolean, Long)] = {
+      if (rangeIDs.isEmpty) addresses
+      else
+        subtractAddressesInRanges(
+          subtractAddressesInRange(addresses, rangeIDs.head, conf, repo),
+          rangeIDs.tail
+        )
+    }
+    subtractAddressesInRanges(getAddressesTuples(leftRangeIDs, conf), rightRangeIDs)
+  }
+
+  /** receives a dataframe containing active and expired ranges and returns addresses contained only in expired ranges
+   *  @param ranges dataframe of type   rangeID:String | expired: Boolean
+   *  @return dataframe of type  key:String | address:String | relative:Boolean | last_modified:Long
+   */
+  def getAddressesDFFromRanges(
+      ranges: Dataset[Row],
+      conf: APIConfigurations,
+      repo: String
+  ): Dataset[Row] = {
+    val left_anti_join_addresses = udf((x: Seq[String], y: Seq[String]) => {
+      leftAntiJoinAddresses(x.toSet, y.toSet, conf, repo).toSeq
+    })
+    val expiredRangesDF = ranges.where("expired")
+    val activeRangesDF = ranges.where("!expired")
+
+    // ranges existing in expired and not in active
+    val uniqueExpiredRangesDF = expiredRangesDF.join(
+      activeRangesDF,
+      expiredRangesDF("range_id") === activeRangesDF("range_id"),
+      "leftanti"
+    )
+
+    // for every expired range - get all intersecting active ranges
+    //  expired_range | active_ranges  | min-max (of expired_range)
```

Comment:
This does _not_ match what I think of when I read "intersecting".  E.g., if u is entirely contained in a then `u.max_key < a.max_key`, so the first comparison on l. 169 is false, and `a.min_key < u.min_key`, so the second comparion on l. 170 is also false.  Is the intent here some kind of "overlapping but not contained"?

Reply:
Thanks, fixing it!